reactor-core/src/main/java/reactor/core/publisher/FluxTransactional.java

```diff
@@ -0,0 +1,653 @@
+/*
+ * Copyright (c) 2011-2018 Pivotal Software Inc, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.core.publisher;
+
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscription;
+import reactor.core.CoreSubscriber;
+import reactor.core.Exceptions;
+import reactor.core.Fuseable;
+import reactor.util.Logger;
+import reactor.util.Loggers;
+import reactor.util.annotation.Nullable;
+
+/**
+ * Uses a resource, generated by a supplier for each individual Subscriber,
+ * while streaming the values from a
+ * Publisher derived from the same resource and makes sure the resource is released
+ * if the sequence terminates or the Subscriber cancels.
+ * <p>
+ * <p>
+ * Eager resource cleanup happens just before the source termination and exceptions
+ * raised by the cleanup Consumer may override the terminal even. Non-eager
+ * cleanup will drop any exception.
+ *
+ * @param <T> the value type streamed
+ * @param <S> the resource type
+ *
+ * @see <a href="https://github.com/reactor/reactive-streams-commons">Reactive-Streams-Commons</a>
+ */
+final class FluxTransactional<T, S> extends Flux<T> implements Fuseable, SourceProducer<T> {
+
+	private static final Logger LOGGER = Loggers.getLogger(FluxTransactional.class);
+
+	final Supplier<S>                                           resourceSupplier;
+	final Function<? super S, ? extends Publisher<? extends T>> transactionClosure;
+	final Function<? super S, ? extends Publisher<?>>           commit;
+	final Function<? super S, ? extends Publisher<?>>           rollback;
+	final boolean                                               rollbackOnCancel;
+
+	FluxTransactional(Supplier<S> transactionalResourceSupplier,
+			Function<? super S, ? extends Publisher<?>> commit,
+			Function<? super S, ? extends Publisher<?>> rollback,
+			Function<? super S, ? extends Publisher<? extends T>> transactionClosure,
+			boolean rollbackOnCancel) {
```

Comment:
replace with a third `Function`, maybe reorder arguments in public API to stick to similar order to using.